fosmount/fosmount.1
.\"
.TH "FOSMOUNT" "1" "January 2025" "fosmount" "User Commands"
.SH "NAME"
fosmount \- FOS mounter for fuse
.SH "SYNOPSIS"
.B fosmount
[\fIoptions\fR] \fIdevice mountpoint\fR
.SH "DESCRIPTION"
FUSE extension for a read\-only access on Smaky FOS.
.TP
\fB\-h\fR \fB\-\-help\fR
Display help message.
.TP
\fB\-v\fR \fB\-\-version\fR
Show version.
.TP
\fB\-a\fR \fB\-\-harddisk\fR
Force an hard disk instead of the autodetection.
.TP
\fB\-f\fR \fB\-\-floppydisk\fR
Force a floppy disk instead of the autodetection.
.TP
\fB\-l\fR \fB\-\-fos\-logger\fR
Turn ON the FOS logger, fosmount will not be detached of the terminal.
.TP
\fB\-d\fR \fB\-\-fuse\-debugger\fR
Turn ON the FUSE debugger, fosmount will not be detached of the terminal.
.TP
\fB\-i\fR \fB\-\-image\-bmp\fR
Convert on the fly .IMAGE and .COLOR to .BMP (The BMP file format).
.TP
\fB\-t\fR \fB\-\-text\fR
Convert on the fly some text files to .TXT (UTF-8).
.TP
\fBdevice\fR
/dev/fd0 for floppy disk
.br
/dev/sda for hard disk, etc, ...
.TP
\fBmountpoint\fR
Mount point for the FOS, like /mnt/smaky.
.SH "AUTHOR"
Written by Mathieu Schroeter <mathieu@schroetersa.ch>.
.SH "REPORTING BUGS"
Report bugs to <\fImathieu@schroetersa.ch\fP>.
.SH "COPYRIGHT"
Copyright \(co 2006\-2010,2025 Mathieu Schroeter

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


ChangeLog
2026-MM-DD  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 2.0.0

	* The Windows version supports binary files (dumps, like with Linux)
	  and not just physical devices.

	* fosmount: replace the -i and -j options by only the -i option with the
	  major difference that BMP files are produced for .IMAGE and .COLOR
	  files instead of PBM and XPM2 which are not supported on Windows.

	* fosmount: add a new -t, --text option useful to convert some text
	  files to the UTF-8 charset.

	* fosread: add recursive copy of a whole directory with the "get" mode.

	* fosread: add -i, --image-bmp and -t, --text options to the fosread
	  command. It enables the possibility to convert these files on Windows
	  like in the case of fosmount for Linux.

	* mosread: now tool to read and extract content of SAMOS disks and
	  images.

2024-10-08  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 1.0.1

	* fosmount: move to fuse3. Thanks to Jakob Haufe from Debian.

2023-03-26  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 1.0.0

	* fosdd: new tool that can be used in order to convert floppy device
	  or image into hard disk image and vice-versa.

	* fosmount: the PBM and XPM images use a shorter filename.

2010-03-14  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Release 0.4.0

	* Many improvements in the build scripts and Makefiles. The build for
	  Windows uses the --cross-compile option with the configure script.
	  The external scripts to build for Windows, provided with the older
	  versions are no longer necessary. Now the DLLs are created in the
	  right way.
	  The pkg-config files are now available for libfosfat and libfosgra.

	* libfosgra: add support for the decoding of .COLOR files.

	* fosmount: add the conversion on the fly of .COLOR files to .XPM files.
	  Note that the XPM files use the version 2 of X PixMap. It needs a real
	  application like The Gimp in order to read the files. Or the tool SXPM
	  from the X11 (X.Org) project.

	* smascii: add a manpage. Thanks to Didier Raboud from Debian.

	* libw32disk: this library doesn't rely anymore to C++. It is now fully
	  rewritten in C with a proper license in order to be redistribuable by
	  Debian.

2009-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.2

	* libfosgra: new library to support .IMAGE decoding.

	* fosmount: some improvements to convert on the fly the images [.IMAGE]
	  to PBM (portable bitmap format) with libfosgra.

	* Fix doxygen, packaging and Makefiles.

2009-07-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.1

2009-05-06  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* fosrec: Add a new tool to easily recovery files.

	* Fix Debian/Ubuntu packaging.

2009-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* A lot of cleaning and all doxygen documentations are only in the
	  public header file.

	* Add option in configure to generate doxygen HTML and PDF outputs.

2008-02-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Improve a lot the code lisibility.

	* Fix some memleaks when errors appear.

2008-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add manpages for fosmount and fosread.

2007-12-16  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New major beta released 0.3.0

	* Add Window$ support, and use MinGW32 to compile DLL libraries,
	  fosread and smascii. A script package-win32.sh can be used to create
	  a zip package.

	* libfosfat: add support for Window$ through libw32disk.

	* libw32disk: add C wrapper to use original Disk.cxx class with
	  libfosfat.

	* fosread: fix symlink support, now that is possible to read like a
	  directory. Add support to list and restore deleted files.

2007-12-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: first implementation of an undelete tool just disabled by
	  default. The next step is added some options to list and restore
	  files.

2007-12-12  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: add support for multiple devices. A new main structure is
	  used instead of global variables. Only the logger is always global.

2007-12-09  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Fourth beta released 0.2.3

	* Some fixes, cleaned code, ...

2007-07-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Change license GPLv2 to GPLv3.

	* Fix a lot of memory leaks.

2007-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Third beta released 0.2.2

2007-06-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: change the auto detection function for a better way. Use
	  the name was very bad because a lot of disks have no name. Then the
	  test is does with the CHK (look the code for more informations).

2007-06-24  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add auto detection for the disk type. The type can be forced with
	  fosmount and fosread if necessary.

	* libfosfat: an internal logger can be enabled to know what the FOS is
	  doing. A new option is available in fosread and fosmount.

2007-06-22  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Second beta released 0.2.1

	* Add support for symbolic links. The Smaky's soft links are now
	  converted to symlink.

	* libfosfat: the cache is now released when the device is closed.

2007-06-20  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: a new system of caching is implemented to improve the
	  speed. Now that will be very fast for looking in a directory. The
	  system must be fixed for that the memory is released.

	* fosmount: a bug is fixed for that FUSE is always loaded as
	  single-thread. If not there is a really a big problem when data are
	  loaded with libfosfat.

2007-05-10  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* First beta released 0.2.0

2007-04-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>


fosmount/fosmount.c
/*
 * FOS fosmount: FUSE extension in read-only for Smaky file system
 * Copyright (C) 2006-2010,2025-2026 Mathieu Schroeter <mathieu@schroetersa.ch>
 *
 * Thanks to Pierre Arnaud for his help and the documentation
 *    And to Epsitec SA for the Smaky computers
 *
 * This file is part of Fosfat.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>     /* strcmp strncmp strstr strlen strdup memcpy memset */
#include <time.h>       /* mktime */
#include <fuse.h>
#include <getopt.h>

#include "fosfat.h"
#include "fosfat_internal.h"
#include "fosgra.h"
#include "utils.h"

/* Rights */
#define FOS_DIR             0555
#define FOS_FILE            0444

#define HELP_DEVICE \
"file.di  : disk image\n" \
"                       /dev/fd0 : floppy disk\n" \
"                       /dev/sda : hard disk, etc\n"

#define HELP_TEXT \
"FUSE extension for a read-only access on Smaky FOS. fosfat-" LIBFOSFAT_VERSION_STR "\n\n" \
"Usage: fosmount [options] device mountpoint\n\n" \
" -h --help             this help\n" \
" -v --version          version\n" \
" -a --harddisk         force an hard disk (default autodetect)\n" \
" -f --floppydisk       force a floppy disk (default autodetect)\n" \
" -l --fos-logger       that will turn on the FOS logger\n" \
" -d --fuse-debugger    that will turn on the FUSE debugger\n" \
" -i --image-bmp        convert on the fly .IMAGE and .COLOR to .BMP\n" \
" -t --text             convert on the fly some text files to .TXT (UTF-8)\n" \
" device                " HELP_DEVICE \
" mountpoint            for example, /mnt/smaky\n" \
"\nPlease, report bugs to <mathieu@schroetersa.ch>.\n"

#define VERSION_TEXT "fosmount-" LIBFOSFAT_VERSION_STR "\n"


static fosfat_t *fosfat;

#define FLYID "@"

static int g_bmp = 0;
static int g_txt = 0;


static char *
trim_fosname (const char *path)
{
  char *it, res[256];

  snprintf (res, sizeof (res), "%s", path);
  it = strstr (res, "." FLYID ".");
  if (it)
    *it = '\0';

  return strdup (res);
}

static size_t
get_filesize (fosfat_file_t *file, const char *path)
{
  fosfat_ftype_t ftype = FOSFAT_FTYPE_OTHER;

  if (file->att.isdir || file->att.islink || file->att.isencoded)
    return file->size;

  ftype = fosfat_ftype (file->name);

  if (g_bmp && ftype == FOSFAT_FTYPE_IMAGE && fosgra_is_image (fosfat, path))
    return fosgra_bmp_get_size (fosfat, path);

  if (g_txt && ftype == FOSFAT_FTYPE_TEXT)
  {
    size_t size = 0;
    uint8_t *text_buffer = fosfat_get_buffer (fosfat, path, 0, file->size);
    uint8_t *buffer = calloc (2, file->size);
    fosfat_sma2utf8 ((const char *) text_buffer, file->size,
                     (char *) buffer, file->size * 2, FOSFAT_ASCII_LF);
    free (text_buffer);
    size = strlen ((char *) buffer);
    free (buffer);
    return size;
  }

  return file->size;
}

static uint8_t *
get_buffer (fosfat_file_t *file, const char *path, off_t offset, size_t size)
{
  fosfat_ftype_t ftype = FOSFAT_FTYPE_OTHER;

  if (file->att.isdir || file->att.islink || file->att.isencoded)
    goto std;

  ftype = fosfat_ftype (file->name);

  if (g_bmp && ftype == FOSFAT_FTYPE_IMAGE && fosgra_is_image (fosfat, path))
  {
    size_t image_size = 0;
    uint8_t *image_buffer = fosgra_bmp_get_buffer (fosfat, path, &image_size);
    uint8_t *dec = calloc (1, size);
    memcpy (dec, image_buffer + offset, size);
    free (image_buffer);
    return dec;
  }

  if (g_txt && ftype == FOSFAT_FTYPE_TEXT)
  {
    uint8_t *text_buffer = fosfat_get_buffer (fosfat, path, 0, file->size);
    uint8_t *dec = calloc (2, file->size);
    uint8_t *buffer = calloc (2, file->size);
    fosfat_sma2utf8 ((const char *) text_buffer, file->size,
                     (char *) buffer, file->size * 2, FOSFAT_ASCII_LF);
    free (text_buffer);
    memcpy (dec, buffer + offset, size);
    free (buffer);
    return dec;
  }

std:
  return fosfat_get_buffer (fosfat, path, offset, size);
}

/*
 * Convert 'fosfat_file_t' to 'struct stat'.
 */
static struct stat *
in_stat (fosfat_file_t *file, const char *path)
{
  struct stat *st;
  struct tm time;

  st = calloc (1, sizeof (struct stat));
  if (!st)
    return NULL;

  memset (&time, 0, sizeof (time));

  /* Directory, symlink or file */
  if (file->att.isdir)
  {
    st->st_mode = S_IFDIR | FOS_DIR;
    st->st_nlink = 2;
  }
  else if (file->att.islink)
  {
    st->st_mode = S_IFLNK | FOS_DIR;
    st->st_nlink = 2;
  }
  else
  {
    st->st_mode = S_IFREG | FOS_FILE;
    st->st_nlink = 1;
  }

  /* Size */
  st->st_size = get_filesize (file, path);

  /* Time */
  time.tm_year = file->time_r.year - 1900;
  time.tm_mon  = file->time_r.month - 1;
  time.tm_mday = file->time_r.day;
  time.tm_hour = file->time_r.hour;
  time.tm_min  = file->time_r.minute;
  time.tm_sec  = file->time_r.second;
  st->st_atime = mktime (&time);
  time.tm_year = file->time_w.year - 1900;
  time.tm_mon  = file->time_w.month - 1;
  time.tm_mday = file->time_w.day;
  time.tm_hour = file->time_w.hour;
  time.tm_min  = file->time_w.minute;
  time.tm_sec  = file->time_w.second;
  st->st_mtime = mktime (&time);
  time.tm_year = file->time_c.year - 1900;
  time.tm_mon  = file->time_c.month - 1;
  time.tm_mday = file->time_c.day;
  time.tm_hour = file->time_c.hour;
  time.tm_min  = file->time_c.minute;
  time.tm_sec  = file->time_c.second;
  st->st_ctime = mktime (&time);

  return st;
}

/*
 * Get the file stat from a path an return as struct stat.
 */
static struct stat *
get_stat (const char *path)
{
  char *location;
  struct stat *st = NULL;
  fosfat_file_t *file;

  location = trim_fosname (path);

  file = fosfat_get_stat (fosfat, location);
  if (file)
  {
    st = in_stat (file, location);
    free (file);
  }

  if (location)
    free (location);

  return st;
}

/*
 * FUSE : read the target of a symlink.
 *
 * path         (foo/bar)
 * dst          target
 * size         max length
 * return 0 for success
 */
static int
fos_readlink (const char *path, char *dst, size_t size)
{
  int res = -ENOENT;
  char *link, *location;

  location = trim_fosname (path);

  link = fosfat_symlink (fosfat, location);

  if (strlen (link) < size)
  {
    memcpy (dst, link, strlen (link) + 1);
    res = 0;
  }

  if (link)
    free (link);
  if (location)
    free (location);

  return res;
}

/*
 * FUSE : get attributes of a file.
 *
 * path         (foo/bar)
 * stbuf        attributes
 * return 0 for success
 */
static int
fos_getattr (const char *path, struct stat *stbuf, struct fuse_file_info *fi)
{
  int ret = 0;
  char *location;
  struct stat *st;

  (void) fi;

  /* Root directory */
  if (!strcmp (path, "/"))
    location = strdup ("/sys_list");
  else
    location = trim_fosname (path);

  /* Get file stats */
  st = get_stat (location);
  if (st)
  {
    memcpy (stbuf, st, sizeof (*stbuf));
    free (st);
  }
  else
    ret = -ENOENT;

  if (location)
    free (location);

  return ret;
}

/*
 * FUSE : read a directory.
 *
 * path         (foo/bar)
 * buf          buffer for filler
 * filler       function for put each entry
 * offset       not used
 * fi           not used
 * return 0 for success
 */
static int
fos_readdir (const char *path, void *buf, fuse_fill_dir_t filler,
             off_t offset, struct fuse_file_info *fi, enum fuse_readdir_flags flags)
{
  int ret = -ENOENT;
  char *location;
  fosfat_file_t *file, *first_file;

  (void) offset;
  (void) fi;
  (void) flags;

  location = trim_fosname (path);

  /* First entries */
  filler (buf, "..", NULL, 0, 0);

  /* Files and directories */
  file = fosfat_list_dir (fosfat, location);
  if (!file)
    goto out;

  first_file = file;

  do
  {
    char _path[256];
    struct stat *st;
    char *name;
    fosfat_ftype_t ftype = FOSFAT_FTYPE_OTHER;

    snprintf (_path, sizeof (_path), "%s/%s", location, file->name);
    remove_dup_slashes (_path);
    st = in_stat (file, _path);

    ftype = fosfat_ftype (file->name);

    /* add identification for .IMAGE and .COLOR translated to BMP */
    if (g_bmp && ftype == FOSFAT_FTYPE_IMAGE && fosgra_is_image (fosfat, _path))
    {
      name = calloc (1, strlen (file->name) + strlen (FLYID) + 6);
      sprintf (name, "%s." FLYID ".bmp", file->name);
    }
    /* add identification for text files translated to ISO-8859-1 */
    else if (g_txt && ftype == FOSFAT_FTYPE_TEXT)
    {
      name = calloc (1, strlen (file->name) + strlen (FLYID) + 6);
      sprintf (name, "%s." FLYID ".txt", file->name);
    }
    else
      name = strdup (file->name);

    if (strstr (name, ".dir"))
      *(name + strlen (name) - 4) = '\0';

    /* Add entry in the file list */
    filler (buf, name, st, 0, 0);
    free (st);
    free (name);
  }
  while ((file = file->next_file));

  fosfat_free_listdir (first_file);
  ret = 0;

 out:
  if (location)
    free (location);

  return ret;
}

/*
 * FUSE : test if a file can be opened.
 *
 * path         (foo/bar)
 * fi           flags
 * return 0 for success
 */
static int
fos_open (const char *path, struct fuse_file_info *fi)
{
  int ret = 0;
  char *location;

  location = trim_fosname (path);

  if ((fi->flags & 3) != O_RDONLY)
    ret = -EACCES;
  else if (!fosfat_isopenexm (fosfat, location))
    ret = -ENOENT;

  if (location)
    free (location);

  return ret;
}

/*
 * FUSE : read the data of a file.
 *
 * path         (foo/bar)
 * buf          buffer for put the data
 * size         size in bytes
 * offset       offset in bytes
 * fi           not used
 * return the size
 */
static int
fos_read (const char *path, char *buf, size_t size,
          off_t offset, struct fuse_file_info *fi)
{
  int res = -ENOENT;
  int length;
  char *location;
  uint8_t *buf_tmp;
  fosfat_file_t *file = NULL;

  (void) fi;

  location = trim_fosname (path);

  /* Get the stats and test if it is a file */
  file = fosfat_get_stat (fosfat, location);
  if (!file)
    goto out;

  if (file->att.isdir)
    goto out;

  length = get_filesize (file, location);

  if (offset < length)
  {
    /* Fix the size in function of the offset */
    if (offset + (signed) size > length)
      size = length - offset;

    /* Read the data */
    buf_tmp = get_buffer (file, location, offset, size);

    /* Copy the data for FUSE */
    if (buf_tmp)
    {
      memcpy (buf, buf_tmp, size);
      free (buf_tmp);
      res = size;
    }
  }
  else
    res = 0;

 out:
  if (file)
    free (file);
  if (location)
    free (location);

  return res;
}

/* Print help. */
void
print_info (void)
{
  printf (HELP_TEXT);
}

/* Print version. */
void
print_version (void)
{
  printf (VERSION_TEXT);
}

/* FUSE implemented functions */
static struct fuse_operations fosfat_oper = {
  .getattr  = fos_getattr,
  .readdir  = fos_readdir,
  .open     = fos_open,
  .read     = fos_read,
  .readlink = fos_readlink,
};

int
main (int argc, char **argv)
{
  int i;
  int next_option;
  int res = 0, fusedebug = 0, foslog = 0;
  char *device;
  char **arg;
  fosfat_disk_t type = FOSFAT_AD;

  const char *const short_options = "adfhlitv";

  const struct option long_options[] = {
    { "harddisk",      no_argument, NULL, 'a' },
    { "fuse-debugger", no_argument, NULL, 'd' },
    { "floppydisk",    no_argument, NULL, 'f' },
    { "help",          no_argument, NULL, 'h' },
    { "fos-logger",    no_argument, NULL, 'l' },
    { "image-bmp",     no_argument, NULL, 'i' },
    { "text",          no_argument, NULL, 't' },
    { "version",       no_argument, NULL, 'v' },
    { NULL,            0,           NULL,  0  }
  };

  /* check options */
  do
  {
    next_option = getopt_long (argc, argv, short_options, long_options, NULL);
    switch (next_option)
    {
    default :           /* unknown */
    case '?':           /* invalid option */
    case 'h':           /* -h or --help */
      print_info ();
      return -1;
    case 'v':           /* -v or --version */
      print_version ();
      return -1;
    case 'f':           /* -f or --floppydisk */
      type = FOSFAT_FD;
      break ;
    case 'a':           /* -a or --harddisk */
      type = FOSFAT_HD;
      break ;
    case 'd':           /* -d or --fuse-debugger */
      fusedebug = 1;
      break ;
    case 'l':           /* -l or --fos-debugger */
      foslog = 1;
      fosfat_logger (1);
      break;
    case 'i':           /* -i or --image-bmp */
      g_bmp = 1;
      break ;
    case 't':           /* -t or --text */
      g_txt = 1;
      break;
    case -1:            /* end */
      break ;
    }
  }
  while (next_option != -1);

  if (argc < optind + 2)
  {
    print_info ();
    return -1;
  }

  /* table for fuse */
  arg = malloc (sizeof (char *) * (3 + fusedebug + foslog));
  if (arg)
  {
    arg[0] = strdup (argv[0]);

    if (fusedebug)
      arg[fusedebug] = strdup ("-d");
    if (foslog)
      arg[fusedebug + foslog] = strdup ("-f");

    device = strdup (argv[optind]);
    arg[1 + fusedebug + foslog] = strdup (argv[optind + 1]);

    /* FUSE must be used as single-thread */
    arg[2 + fusedebug + foslog] = strdup ("-s");
  }
  else
    return -1;

  /* Open the floppy disk (or hard disk) */
  fosfat = fosfat_open (device, type, 0);
  if (!fosfat)
  {
    fprintf (stderr, "Could not open %s for mounting!\n", device);
    res = -1;
  }
  else
  {
    /* FUSE */
    res = fuse_main (3 + fusedebug + foslog, arg, &fosfat_oper, NULL);

    /* Close the device */
    fosfat_close (fosfat);
  }

  for (i = 0; i < 3 + fusedebug + foslog; i++)
    free (*(arg + i));
  free (arg);

  /* Free */
  free (device);

  return res;
}
